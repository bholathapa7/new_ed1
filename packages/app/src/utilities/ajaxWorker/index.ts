import { Observable, from } from 'rxjs';
import * as RxjsAjax from 'rxjs/ajax';
import { AjaxCreationMethod } from 'rxjs/internal/observable/dom/AjaxObservable';
import Worker from 'worker-loader!./worker';
import { AJAX_METHOD, AjaxMethod, AjaxParams } from './types';

function createAjaxFromWorker(w: Worker, method: AjaxMethod): Observable<AjaxResponse> {
  return from(new Promise<AjaxResponse>((resolve, reject) => {
    w.onmessage = ({ data }: MessageEvent) => {
      const resp: AjaxResponse = method === AJAX_METHOD.RAW ? data : JSON.parse(data);

      resolve(resp);
      w.terminate();
    };
    w.onerror = (msgEvent: ErrorEvent) => {
      reject(JSON.parse(msgEvent.error));
      w.terminate();
    };
  }));
}


function initializeWorker(ajaxParams: AjaxParams | (string | AjaxRequest), method: AjaxMethod): Observable<AjaxResponse> {
  const worker: Worker = new Worker();
  worker.postMessage(JSON.stringify({ ajaxParams, method }));

  return createAjaxFromWorker(worker, method);
}

export const ajax: {
  get: AjaxCreationMethod['get'];
  post: AjaxCreationMethod['post'];
  delete: AjaxCreationMethod['delete'];
  patch: AjaxCreationMethod['patch'];
  basic(urlOrRequest: string | AjaxRequest): Observable<AjaxResponse>;
  raw(urlOrRequest: string | AjaxRequest): Observable<AjaxResponse>;
} = {
  raw: (urlOrRequest) => initializeWorker(urlOrRequest, AJAX_METHOD.RAW),
  basic: (urlOrRequest) => initializeWorker(urlOrRequest, AJAX_METHOD.BASIC),
  get: (url, headers) => initializeWorker({ url, headers }, AJAX_METHOD.GET),
  post: (url, body, headers) => initializeWorker({ url, headers, body }, AJAX_METHOD.POST),
  delete: (url, headers) => initializeWorker({ url, headers }, AJAX_METHOD.DELETE),
  patch: (url, body, headers) => initializeWorker({ url, headers, body }, AJAX_METHOD.PATCH),
};

export type AjaxError = RxjsAjax.AjaxError;
export type AjaxRequest = RxjsAjax.AjaxRequest;
export type AjaxResponse = RxjsAjax.AjaxResponse;
