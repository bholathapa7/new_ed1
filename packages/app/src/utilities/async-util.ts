import * as Sentry from '@sentry/browser';

export interface TryCatchOutput<T> {
  data: T | null;
  error: Error | null;
}

/**
 * Abstraction to make async actions easier.
 */
export async function tryCatch<T>(promise: Promise<T>): Promise<TryCatchOutput<T>> {
  try {
    const data: T = await promise;

    return { data, error: null };
  } catch (error) {
    Sentry.captureException(error);

    return { error, data: null };
  }
}
