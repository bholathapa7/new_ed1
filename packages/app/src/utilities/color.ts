import Color from 'color';

export const isTransparent: (color: Color) => boolean = (color) => color.alpha() === 0;
export const isWhite: (color: Color) => boolean = (color) => color.luminosity() === 1;
