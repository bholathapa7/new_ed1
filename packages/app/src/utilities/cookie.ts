import * as T from '^/types';

const tail: string = 'path=/; domain=.angelswing.io;';

export const getCookie: (key: string) => string | undefined = (key) => {
  const cookies: string[] = `; ${document.cookie}`.split(`; ${key}=`);

  if (cookies.length === 2) return cookies.pop()?.split(';').shift();

  return;
};

/**
 * @desc This function is for setting cookie. Receives key & value.
 */
export const setCookie: (key: string, value: string, expiredAt?: Date) => void = (
  key, value, date,
) => {
  const expiredAt: string = date ? `expires = ${date.toUTCString()};` : '';
  document.cookie = `${key}=${value}; ${expiredAt} ${tail}`;
};

/**
 * @desc This function is for setting cookie with Object. Receives object.
 */
export const setCookieWithObject: (obj: { [K: string]: string }, expiredAt?: Date) => void = (
  obj, expiredAt,
) => Object.keys(obj).forEach((key: string) => setCookie(key, obj[key], expiredAt));

/**
 * @desc This function is for deleting cookie. Receives key or keys.
 */
export const deleteCookie: (...key: Array<string>) => void = (...key) =>
  key.forEach((k: string) => document.cookie = `${k}= ; ${tail} expires=Thu, 01 Jan 1970 00:00:00 GMT`);

/**
 * This function is for setting cloudfront cookie. Receives authedUser in AuthState.
 */
export const setCloudFrontCookie: (cloudFront: T.CloudFront) => void = ({
  policy, signature, keyPairId,
}) => {
  const cookie: T.CookieCloudFront = {
    'CloudFront-Policy': policy,
    'CloudFront-Signature': signature,
    'CloudFront-Key-Pair-Id': keyPairId,
  };
  const expiredAt: Date = new Date();
  expiredAt.setDate(expiredAt.getDate() + 1);

  setCookieWithObject({ ...cookie }, expiredAt); // Use spread for Type Error
};

export const deleteCloudFrontCookie: () => void = () => {
  const keys: Array<keyof T.CookieCloudFront> = ['CloudFront-Key-Pair-Id', 'CloudFront-Policy', 'CloudFront-Signature'];
  deleteCookie(...keys);
};
