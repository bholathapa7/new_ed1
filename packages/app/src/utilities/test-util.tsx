import { createLocation } from 'history';
import createMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store';
import { action as makeAction, payload, union } from 'tsdux';

import * as T from '^/types';

import * as M from '^/store/Mock';

const mockStoreCreator: MockStoreCreator<any> = createMockStore<any>();

/**
 * @author Joon-Mo Yang <jmyang@angelswing.io>
 * @desc mock context creator for common page state
 */

export function mockCommonContext() {
  const commonStore = {
    Pages: {
      Common: M.mockCommonPage,
    },
  };

  return {
    store: (mockStoreCreator as MockStoreCreator<typeof commonStore>)(commonStore),
  };
}


export const UpdateLanguage = makeAction(
  'test/UPDATE_LANGUAGE',
  payload<T.Language>(),
);

export const Auth = makeAction(
  'test/AUTH',
);
export const Unauth = makeAction(
  'test/UNAUTH',
);

const Action = union([
  UpdateLanguage,
  Auth,
  Unauth,
]);
export type DDMMockAction = typeof Action;


export type DDMMockStore = MockStoreEnhanced<T.State>;

export const initialMockState: T.State = {
  Attachments: M.mockAttachments,
  Auth: M.mockAuth,
  Contents: M.mockContents,
  Screens: M.mockScreens,
  Pages: M.mockPages,
  Permissions: M.mockPermissions,
  Projects: M.mockProjects,
  SharedContents: M.mockSharedContents,
  Users: M.mockUsers,
  UserConfig: M.mockUserConfig,
  PlanConfig: M.mockPlanConfig,
  ESSModels: M.mockESSModels,
  ESSContents: M.mockESSContents,
  ESSAttachments: M.mockESSAttachments,
  ProjectConfigPerUser: M.mockProjectConfigPerUser,
  Photos: M.mockPhotos,
  Groups: M.mockGroups,
  router: {
    location: createLocation(''),
    action: 'PUSH',
  },
};

export const createDDMMockStore: (
  initialLanguage: T.Language,
) => DDMMockStore = (
  initialLanguage,
) => {
  const initialState: T.State = {
    ...initialMockState,
    Pages: {
      ...M.mockPages,
      Common: {
        ...M.mockCommonPage,
        language: initialLanguage,
      },
    },
    Permissions: M.mockPermissions,
    Projects: M.mockProjects,
    SharedContents: M.mockSharedContents,
    Users: M.mockUsers,
    router: {
      location: createLocation(''),
      action: 'PUSH',
    },
    ProjectConfigPerUser: {
      config: undefined,
      patchProjectConfigStatus: T.APIStatus.IDLE,
    },
    PlanConfig: M.mockPlanConfig,
  };

  return mockStoreCreator((
    actions: Array<DDMMockAction>,
  ) => actions.reduce((state, action) => {
    switch (action.type) {
      case UpdateLanguage.type:
        return {
          ...state,
          Pages: {
            ...state.Pages,
            Common: {
              ...state.Pages.Common,
              language: action.payload,
            },
          },
        };
      case Auth.type:
        return {
          ...state,
          Auth: {
            ...state.Auth,
            authedUser: {
              id: state.Users.users.allIds[0],
              token: 'random-test-token',
            },
          },
        };
      case Unauth.type:
        return {
          ...state,
          Auth: {
            ...state.Auth,
            authedUser: undefined,
          },
        };
      default:
        return state;
    }
  }, initialState));
};

/**
 * @desc common `afterEach` hook for mount testing
 */
export function commonAfterEach(): void {
  // @ts-ignore
  jest.clearAllTimers();
  // @ts-ignore
  jest.resetAllMocks();
}
