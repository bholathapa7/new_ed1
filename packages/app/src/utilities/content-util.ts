import { isSameDay } from 'date-fns';
import { isLonLat, typeGuardGroup } from '^/hooks';
import { TabToCategoryMapper } from '^/store/duck/Groups';
import { getCenterBoundary, getMidPoint } from '^/utilities/map-util';

import * as T from '^/types';
import { L10nDictionary } from './l10n';
import proj4 from 'proj4';
import { getEPSGfromProjectionLabel } from './coordinate-util';
import { gcpsToGeoPoints } from './gcp-util';

export const isPinnable: (category: T.Content['category']) => boolean = (category) => category === T.ContentCategory.MEASUREMENT;
export const isContentPinned: (content: T.Content) => boolean = (content) => content.screenId === undefined;

/**
 * @todo: should be changed after guideline comes out
 */
export const contentTexts: { [K in T.ContentType]: L10nDictionary } = {
  [T.ContentType.AREA]: {
    [T.Language.KO_KR]: '면적',
    [T.Language.EN_US]: 'Area',
  },
  [T.ContentType.LENGTH]: {
    [T.Language.KO_KR]: '거리',
    [T.Language.EN_US]: 'Distance',
  },
  [T.ContentType.MARKER]: {
    [T.Language.KO_KR]: '위치',
    [T.Language.EN_US]: 'Location',
  },
  [T.ContentType.VOLUME]: {
    [T.Language.KO_KR]: '체적',
    [T.Language.EN_US]: 'Volume',
  },
  [T.ContentType.BLUEPRINT_PDF]: {
    [T.Language.KO_KR]: '도면 (.PDF)',
    [T.Language.EN_US]: 'PDF',
  },
  [T.ContentType.BLUEPRINT_DXF]: {
    [T.Language.KO_KR]: '도면 (.DXF)',
    [T.Language.EN_US]: 'DXF',
  },
  [T.ContentType.BLUEPRINT_DWG]: {
    [T.Language.KO_KR]: '도면 (.DWG)',
    [T.Language.EN_US]: 'DWG',
  },
  [T.ContentType.DESIGN_DXF]: {
    [T.Language.KO_KR]: '계획고 (.DXF)',
    [T.Language.EN_US]: 'DXF',
  },
  [T.ContentType.MAP]: {
    [T.Language.KO_KR]: '정사영상',
    [T.Language.EN_US]: '2D Orthomosaic',
  },
  [T.ContentType.DSM]: {
    [T.Language.KO_KR]: '수치표면모델',
    [T.Language.EN_US]: 'DSM',
  },
  [T.ContentType.POINTCLOUD]: {
    [T.Language.KO_KR]: '포인트 클라우드',
    [T.Language.EN_US]: 'Point Cloud',
  },
  [T.ContentType.THREE_D_ORTHO]: {
    [T.Language.KO_KR]: '3D 정사영상',
    [T.Language.EN_US]: '3D Orthomosaic',
  },
  [T.ContentType.THREE_D_MESH]: {
    [T.Language.KO_KR]: '3D 메쉬 모델',
    [T.Language.EN_US]: '3D Mesh Model',
  },
  [T.ContentType.GCP_GROUP]: {
    [T.Language.KO_KR]: '지상기준점',
    [T.Language.EN_US]: 'GCP',
  },
  [T.ContentType.ESS_MODEL]: {
    [T.Language.KO_KR]: '장비',
    [T.Language.EN_US]: 'Equipment',
  },
  [T.ContentType.ESS_ARROW]: {
    [T.Language.KO_KR]: '화살표',
    [T.Language.EN_US]: 'Arrow',
  },
  [T.ContentType.ESS_POLYGON]: {
    [T.Language.KO_KR]: '다각형',
    [T.Language.EN_US]: 'Polygon',
  },
  [T.ContentType.ESS_POLYLINE]: {
    [T.Language.KO_KR]: '선',
    [T.Language.EN_US]: 'Polyline',
  },
  [T.ContentType.ESS_TEXT]: {
    [T.Language.KO_KR]: '텍스트',
    [T.Language.EN_US]: 'Text',
  },
  [T.ContentType.GROUP]: {
    [T.Language.KO_KR]: '그룹 제목',
    [T.Language.EN_US]: 'Group Title',
  },
};

/**
 * @desc This function returns array of content titles about pinned, unpinned contents in specific date
 */
export const getMeasurementContentTitlesFromDate: (
  contents: T.ContentsState['contents']['byId'], date?: Date,
) => Array<string> = (contents, date) => Object
  .values(contents)
  .filter((content: T.Content) => T.MeasurementContentTypes.includes(content.type))
  .filter(({ appearAt }: T.MeasurementContent) => appearAt === undefined || (date !== undefined && isSameDay(appearAt, date)))
  .map(({ title }) => title);

export const getContentTitlesByType: (
  contents: T.ContentsState['contents']['byId'], type: T.ESSContent['type'],
) => string[] = (contents, type) => Object
  .values(contents)
  .reduce<string[]>((names, content) => {
    if (content.type === type) {
      names.push(content.title);
    }

    return names;
  }, []);

/**
 * Gets the first occurence of a group given the category.
 * This is used when no group id is selected.
 * If screen id is provided, it will go through ALL contents from category
 * if there is none, only then use the first pinned group.
 * When no screen id provided, it will use the first pinned group regardless.
 *
 * @param category All categories that have groups, i.e. measurements, overlay, ess
 * @param screenId
 * @returns The group id. If undefined is returned, this means this project has zero groups.
 */
const getFirstGroupByCategory: (
  category: T.ContentCategory, screenId?: T.Screen['id'],
) => (state: T.State) => T.GroupContent['id'] | undefined = (
  category, screenId,
) => (state) => {
  let pinnedGroup: T.GroupContent['id'] | undefined;
  let unpinnedGroup: T.GroupContent['id'] | undefined;

  for (const id of state.Contents.contents.allIds) {
    const content: T.Content | undefined = state.Contents.contents.byId[id];
    if (content === undefined || content.type !== T.ContentType.GROUP || content.category !== category) {
      continue;
    }

    pinnedGroup = content.id;
    if (screenId === undefined) {
      break;
    }

    if (screenId === content.screenId) {
      unpinnedGroup = content.id;
      break;
    }
  }

  return unpinnedGroup ?? pinnedGroup;
};

export const getCurrentGroupId: (
  state: T.State,
  tab: T.ContentPageTabType,
  lastSelectedScreenId?: T.Screen['id'],
) => T.GroupContent['id'] = (
  state, tab, lastSelectedScreenId
) => {
  const selectedGroupId = state.Groups.selectedGroupIdByTab[tab];
  const groupContent = typeGuardGroup(state.Contents.contents.byId[selectedGroupId ?? NaN]);
  // There's a small chance that somehow the group does not exist.
  // If so, fallback to at least one group from the list below.
  if (groupContent !== undefined) {
    return groupContent.id;
  }

  const firstGroupId = getFirstGroupByCategory(TabToCategoryMapper[tab], lastSelectedScreenId)(state);
  if (firstGroupId === undefined) {
    throw new Error('Unable to get the first group: at least one group should exist in the project to create a content.');
  }

  return firstGroupId;
};

/**
 * Get center point of any content.
 */
export const getCenterPointByContent: (
  content: T.Content, projectProjection?: T.ProjectionEnum,
) => T.GeoPoint | undefined = (
  content, projectProjection
) => {
  switch (content.type) {
    case T.ContentType.BLUEPRINT_PDF:
      return getCenterBoundary({
        minLon: content.info.geoPoint[0][0],
        maxLon: content.info.geoPoint[1][0],
        minLat: content.info.geoPoint[0][1],
        maxLat: content.info.geoPoint[1][1],
      });
    case T.ContentType.BLUEPRINT_DXF:
    case T.ContentType.BLUEPRINT_DWG:
      if (content.info.tms?.center) {
        return [content.info.tms.center[0], content.info.tms.center[1]];
      }
      if (content?.info.tms?.bounds) {
        return getCenterBoundary({
          minLon: content.info.tms.bounds[T.MapBounds.MIN_LON],
          maxLon: content.info.tms.bounds[T.MapBounds.MAX_LON],
          minLat: content.info.tms.bounds[T.MapBounds.MIN_LAT],
          maxLat: content.info.tms.bounds[T.MapBounds.MAX_LAT],
        });
      }

      return undefined;
    case T.ContentType.ESS_MODEL:
    case T.ContentType.ESS_TEXT:
    case T.ContentType.MARKER:
      if (projectProjection !== undefined) {
        return isLonLat(content.info.location)
          ? content.info.location
          : proj4(getEPSGfromProjectionLabel(projectProjection), 'EPSG:4326').forward(content.info.location);
      }

      return undefined;
    case T.ContentType.ESS_ARROW:
    case T.ContentType.ESS_POLYGON:
    case T.ContentType.ESS_POLYLINE:
    case T.ContentType.LENGTH:
    case T.ContentType.AREA:
    case T.ContentType.VOLUME:
      return getMidPoint(content.info.locations);
    case T.ContentType.GCP_GROUP:
      const geoPoints: T.GeoPoint[] | undefined = gcpsToGeoPoints(content.info.gcps, content.info.crs);
      if (geoPoints !== undefined) {
        return getMidPoint(geoPoints);
      }

      return undefined;
    default:
      return undefined;
  }
};
