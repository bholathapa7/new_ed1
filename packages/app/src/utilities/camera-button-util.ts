import download from '^/utilities/download';
import html2canvas from 'html2canvas';

import * as T from '^/types';
import { TryCatchOutput, tryCatch } from '^/utilities/async-util';
import { getUserAgent } from './userAgent';

export const runScreenCapture: (element: HTMLElement, capturedFileName: string) => Promise<void> =
  async (element, capturedFileName) => {
    /**
     * @description When browser is webkit based
     * and usePopup parameter of download function is false, then the browser redirects to blob url.
     */
    const isWebkit: boolean = getUserAgent() === T.UserAgent.SAFARI;

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    const { data, error }: TryCatchOutput<HTMLCanvasElement> = await tryCatch(html2canvas(element, {
      allowTaint: true,
      useCORS: true,
    }) as Promise<HTMLCanvasElement>);

    if (!error && data) {
      download(data.toDataURL(), isWebkit, true, capturedFileName);
    }
  };

interface TwoDHandleCameraClickParams {
  onStart(): void;
  onEnd(): void;
  mainScreenCaptureFunction(): Promise<void>;
}

type TwoDHandleCameraClick = ({
  onStart,
  onEnd,
  mainScreenCaptureFunction,
}: TwoDHandleCameraClickParams) => () => Promise<void>;

export const handleCameraClickFunctor: TwoDHandleCameraClick = ({
  onStart,
  onEnd,
  mainScreenCaptureFunction,
}) => async () => {
  onStart();
  await mainScreenCaptureFunction();
  onEnd();
};
