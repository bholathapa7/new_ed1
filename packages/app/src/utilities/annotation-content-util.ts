import * as _ from 'lodash-es';
import { Coordinate } from 'ol/coordinate';

import { CONTENT_NUMBER_GAP, CONTENT_TITLE_PREFIX, START_CONTENT_NUMBER, START_MEASUREMENT_NUMBER } from '^/constants/defaultContent';
import { L10nFn } from '^/hooks';
import * as T from '^/types';
import { contentTexts } from './content-util';

export const getOrderedTitle: (name: T.Content['title'], usingNames: Array<T.Content['title']>) => T.Content['title'] = (
  name, usingNames,
) => {
  const postfix: Array<number> = usingNames
    // Contents must start with the exact name
    .filter((n) => n.includes(name) && n.indexOf(name) === 0)
    // Only retrieve anything else after the name to increment the number.
    .map((n) => n.replace(name, '').match(/\s*\d*$/))
    .map((m) => m === null ? START_MEASUREMENT_NUMBER : Number(m[0]));
  const usingOrdered: Array<number> = _.uniq(postfix);
  const maxOrder: number | undefined = _.max(usingOrdered);

  /*
   * I tried to start newOrder from 1 and not display 1 in a way.
   * However, if the user changes to 1 directly, it starts from 0 because it needs to branch.
   */
  let newOrder: number = START_MEASUREMENT_NUMBER;
  while (maxOrder !== undefined && newOrder <= maxOrder && usingOrdered.includes(newOrder)) {
    if (newOrder === START_MEASUREMENT_NUMBER) {
      newOrder++;
    }

    newOrder++;
  }

  return usingOrdered.length && newOrder > START_MEASUREMENT_NUMBER ? `${name} ${newOrder}` : name;
};

export const getCopiedContentTitle: (
  selectedContentTitle: T.Content['title'],
  usingContentTitles: Array<T.Content['title']>
) => T.Content['title'] = (
  selectedContentTitle,
  usingContentTitles,
) => {
  const postfixOfContentsWithSameType: Array<number> = usingContentTitles
    .filter((title) => title.includes(selectedContentTitle) && title.indexOf(selectedContentTitle) === 0)
    .map((title) => {
      const postfix = title.replace(selectedContentTitle, '');
      const parsedNum = Number(postfix.match(/\(([^)]+)\)/)?.pop()) ?? START_CONTENT_NUMBER;
      return parsedNum;
    })
    .filter((num) => !isNaN(num))
    .sort((a, b) => a - b);
  const nextContentNum: number = postfixOfContentsWithSameType.length
    ? postfixOfContentsWithSameType[postfixOfContentsWithSameType.length - 1] + CONTENT_NUMBER_GAP
    : START_CONTENT_NUMBER + CONTENT_NUMBER_GAP;

  return `${selectedContentTitle}${CONTENT_TITLE_PREFIX}(${nextContentNum})`;
};

export const getNotDuplicatePoints: (locations: Array<Coordinate>) => Array<Coordinate> = (
  locations,
) => {
  const refinedLocations: Array<Coordinate> = [];
  locations
    .map((p) => p.toString())
    .forEach((hash, i) => {
      const lastPoint: Coordinate | undefined = _.last(refinedLocations);
      if (lastPoint && lastPoint.toString() === hash) {
        return;
      }
      refinedLocations.push(locations[i]);
    });

  return refinedLocations;
};

export const getContentTitle: (content: T.Content, l10n: L10nFn) => T.Content['title'] = (content, l10n) => {
  switch (content.type) {
    case T.ContentType.MAP:
    case T.ContentType.DSM:
    case T.ContentType.POINTCLOUD:
    case T.ContentType.THREE_D_ORTHO:
    case T.ContentType.GCP_GROUP:
    case T.ContentType.THREE_D_MESH:
      return l10n(contentTexts[content.type]);

    default:
      return content.title;
  }
};

export const getHasColor: (contentType: T.ContentType) => boolean = (
  contentType,
) => [
  T.ContentType.MARKER,
  T.ContentType.LENGTH,
  T.ContentType.AREA,
  T.ContentType.VOLUME,
  T.ContentType.ESS_ARROW,
  T.ContentType.ESS_POLYLINE,
  T.ContentType.ESS_POLYGON,
].includes(contentType);
