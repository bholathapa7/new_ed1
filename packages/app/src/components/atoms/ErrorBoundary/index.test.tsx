import * as ReactTesting from '@testing-library/react';
import _ from 'lodash-es';
import React, { FC } from 'react';

import { commonAfterEach } from '^/utilities/test-util';

import { ErrorBoundary } from '.';

const TestComp: FC<{ error?: boolean }> = ({ error }) => {
  if (!error) {
    return (
      <>
        Normal Text
      </>
    );
  } else {
    throw new Error('hi');
  }
};

describe('ErrorBoundary', () => {
  let renderResult: ReactTesting.RenderResult;

  beforeEach(() => {
    renderResult = ReactTesting.render(
      <ErrorBoundary>
        <TestComp />
      </ErrorBoundary>,
    );
  });

  afterEach(ReactTesting.cleanup);
  afterEach(commonAfterEach);

  it('should match with the snapshot', () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should render its children when there is no error', () => {
    expect(renderResult.queryByText('Normal Text')).toBeTruthy();
  });

  it('should render error display when an error occurs', () => {
    jest.spyOn(console, 'error').mockImplementation(_.noop);

    try {
      renderResult.rerender(
        <ErrorBoundary>
          <TestComp error={true} />
        </ErrorBoundary>,
      );
    } catch {
      expect(renderResult.asFragment).toMatchSnapshot();
    }
  });
});
