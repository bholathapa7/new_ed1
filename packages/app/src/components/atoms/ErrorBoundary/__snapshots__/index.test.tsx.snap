// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should match with the snapshot 1`] = `
<DocumentFragment>
  Normal Text
</DocumentFragment>
`;

exports[`ErrorBoundary should render error display when an error occurs 1`] = `[Function]`;
